### PR TITLE
ENH: Use numpy.polynomial for linear fit instead of statsmodels

### LINKED ISSUE #157: Use Polynomial.fit instead of statsmodels for linear regression in Moore's law tutorial
I propose to replace the `statsmodels` usage with NumPy fitting functionality in the Moore's law tutorial. There's nothing wrong with the current approach, but this is a nice opportunity to highlight the fitting functionality within numpy itself for such a simple model.

### PR DESCRIPTION
**Pull request by rossbar**
This replaces the usage of `statsmodels` in the Moore's law tutorial with numpy.polynomial.Polynomial.fit, as this is a nice opportunity to demonstrate the Polynomial package fitting. This has the added benefit of reducing the dependency footprint and hopefully working around the latest CI failures for Python 3.11.

One downside of switching from `sm.OLS` to `Polynomial.fit` is the latter does not return the covariance matrix. IMO the benefits of demonstrating polynomial fitting with NumPy outweigh this, and methods for estimating the standard error of model parameters can be added back in a follow-up PR.

Closes #157

### CODE REVIEW

File: content/mooreslaw-tutorial.md
Comment (bsipocz) [inline content/mooreslaw-tutorial.md:45]: unrelated to the PR, but it would be nice to also link numpy? (looks a bit weird in the rendered version that only mpl is linked)

File: content/mooreslaw-tutorial.md
Comment (bsipocz) [inline content/mooreslaw-tutorial.md:417]: unrelated again, and an extreme nitpick, but the notes strings in the cell above look a bit stupid. The cells are long and having the sidebar, yet `B, A` and the `)` are rendered in a practically empty line.

**Review by bsipocz — APPROVED** [review]
All looks good to me. Thanks!

File: content/mooreslaw-tutorial.md
Comment (rossbar) [reply content/mooreslaw-tutorial.md:417]: Thanks for the feedback - I think there are definitely other changes we could consider here. Attaching metadata to data files to describe the underlying data is a good practice, but I'm not sure that it's something you'd necessarily want to advertise doing with npz files - there are much more full-featured file formats with built-in support for metadata (e.g. HDF5).

Either way, I intend (and encourage others) to make such suggestions, but will do so in separate PRs so this one can stay focused on the `statsmodels` -> `np.polynomial` transition!

**Comment by rossbar** [comment]
@cooperrc any thoughts on these proposed changes? It'd be great if you could double-check that the proposed changes haven't completely butchered the intended narrative!

**Comment by cooperrc** [comment]
> @cooperrc any thoughts on these proposed changes? It'd be great if you could double-check that the proposed changes haven't completely butchered the intended narrative!

Thanks @rossbar, I don't see any issues with narrative. 

I went back-and-forth on what method to use for curve-fitting for this tutorial. The final decision to use `statsmodels` was based on conversations on best practices for building this kind of model. 

As a NumPy tutorial, I see the benefit of keeping it focused on NumPy functions. I do miss the error bars on the best fit results, but that may be my academic engineering background.

**Comment by rossbar** [comment]
> I do miss the error bars on the best fit results, but that may be my academic engineering background.

Agreed - this should be added back in. Unfortunately the `Polynomial.fit` method doesn't provide a way to do this calculation automatically like `np.polyfit` with `cov=True`. This is a commonly requested feature but AFAIK it hasn't been added yet.

However IMV this another opportunity to improve the tutorial by demonstrating how to compute the covariance matrix of the coefficient estimates.

My vote would be to leave that as a follow-up PR so that it doesn't block this one (assuming everything else looks good)

**Comment by bsipocz** [comment]
@rossbar - how do you feel about moving forward with this? Besides the pedagogical reasons, removing statsmodels would be also great for managing our dependencies (e.g. currently they block us using numpy 2.0.dev here...)

**Pull request merged**
